### PR TITLE
[SPARK-49665][SQL] Trim collation support for string functions

### PR DESCRIPTION
**Pull request by jovanpavl-db**
### What changes were proposed in this pull request?
Implement trim support for string functions. Not all functions are covered in this pr, follow up: https://issues.apache.org/jira/browse/SPARK-50012

### Why are the changes needed?
Implement trim support for string functions. Not all functions are covered in this pr, follow up: 

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Tests added in CollationStringExpressionsSuite.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 1]
```diff
@@ -1193,6 +1193,10 @@ public static String[] getICULocaleNames() {
     return Collation.CollationSpecICU.ICULocaleNames;
   }
 
+  public static UTF8String applyTrimmingPolicy(UTF8String input, int collationId) {
+    return Collation.CollationSpec.applyTrimmingPolicy(input, collationId);
+  }
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:1198]: perhaps we should reconsider redirecting a 1-line method call

no hard opinion, but doesn't seem useful at first glance

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: should we go ahead and implement them all here in one sweep? it'll get a bit clunky, but it's small changes anyway - should be easier to review all at once

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -827,7 +844,8 @@ case class ValidateUTF8(input: Expression) extends RuntimeReplaceable with Impli
     Seq(input),
     inputTypes)
 
-  override def inputTypes: Seq[AbstractDataType] = Seq(StringTypeWithCaseAccentSensitivity)
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(StringTypeWithCaseAccentSensitivity)
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:848]: either drop a comment / TODO note, or undo the change

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 3]
```diff
@@ -796,6 +812,7 @@ case class MakeValidUTF8(input: Expression) extends RuntimeReplaceable with Impl
 
 }
 
+// TODO: fix
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:815]: usually in the Spark repo, we tend to tag TODOs using a corresponding Jira ticket, our your github username

the point is that we make TODOs less random and provide concrete references

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java (hunk 1 above)
Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:1198]: In this case I had to because CollationSpec is not visible (private class) outside of the CollationFactory.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 3 above)
Comment (jovanpavl-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:815]: Sorry, should have put WIP as I should clean-up this pr a bit before review..

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java
Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: Sure we can also do that. I chose this set of functions as they were needed for string functions.

Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: Now that I check they are all implemented here.

Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: I think it would be useful to add a brief comment explaining "unawareness" with respect to trimming sensitivity for the remaining expressions implemented as part of `CollationSupport`

for example:
```
  public static class Upper {
    public static UTF8String exec(final UTF8String v, final int collationId, boolean useICU) {
      CollationFactory.Collation collation = CollationFactory.fetchCollation(collationId);
      // Space trimming does not affect the output of this expression.
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 4]
```diff
@@ -46,14 +49,11 @@ public static UTF8String[] exec(final UTF8String s, final UTF8String d, final in
       }
     }
     public static String genCode(final String s, final String d, final int collationId) {
-      CollationFactory.Collation collation = CollationFactory.fetchCollation(collationId);
       String expr = "CollationSupport.StringSplitSQL.exec";
-      if (collation.isUtf8BinaryType) {
+      if (CollationFactory.UTF8_BINARY_COLLATION_ID == collationId) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:53]:
```suggestion
      if (collationId == CollationFactory.UTF8_BINARY_COLLATION_ID) {
```

(or update the others - pick according to your preference, but let's keep it consistent across the file)

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 5]
```diff
@@ -295,17 +299,15 @@ public static UTF8String execICU(final UTF8String v, final int collationId) {
 
   public static class FindInSet {
     public static int exec(final UTF8String word, final UTF8String set, final int collationId) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:301]: a comment would be useful for developers, explaining that the trimming sensitivity is taken into account "under the hood" because the collation-aware `FindInSet` implementation relies on the collator's comparator function

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 6]
```diff
@@ -324,10 +354,16 @@ class CollationStringExpressionsSuite
     case class FindInSetTestCase[R](left: String, right: String, collation: String, result: R)
     val testCases = Seq(
       FindInSetTestCase("AB", "abc,b,ab,c,def", "UTF8_BINARY", 0),
+      FindInSetTestCase("b ", "abc,b,ab,c,def", "UTF8_BINARY_RTRIM", 2),
```
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:357]: please add:
`      FindInSetTestCase("def", "abc,b,ab,c,def ", "UTF8_BINARY_RTRIM", 2),`

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 7]
```diff
@@ -500,9 +502,12 @@ public static UTF8String exec(final UTF8String srcString) {
     }
     public static UTF8String exec(
         final UTF8String srcString,
-        final UTF8String trimString,
+        UTF8String trimString,
         final int collationId) {
       CollationFactory.Collation collation = CollationFactory.fetchCollation(collationId);
+      if(collation.supportsSpaceTrimming) {
+        trimString = CollationFactory.applyTrimmingPolicy(trimString, collationId);
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:509]: I don't think this is the behaviour that we want - iiuc, you're trimming away spaces from the trimString (i.e. the users specifies that they want to trim spaces, and you revoke that given the collation spec - which is incorrect)

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:509]: note that all of these:

- `trim(..., "xy")`
- `trim(..., "xy ")`
- `trim(..., "x y")`
- `trim(..., " xy")`

should generally return exactly the same result (regardless of the collation, as long as they all use the same collation of course)

Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:509]: Oh I see, you are right..

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:509]: so here are some examples to sum it all up:
- `trim("abcxyyxx", "xy")` -> should return: "abc" // suppose: UTF8_BINARY
- `trim("abcxyyxx    ", "xy")` -> should return: "abc    " // suppose: UTF8_BINARY_RTRIM
- `trim("abcxyyxx    ", "xy ")` -> should return: "abc" // suppose: UTF8_BINARY
- `trim("abcxyyxx    ", "xy ")` -> should return: "abc" // suppose: UTF8_BINARY_RTRIM

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 8]
```diff
@@ -556,9 +558,12 @@ public static UTF8String exec(final UTF8String srcString) {
...
+      }
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:566]: of course, same as the above goes for `StringTrimLeft`

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 9]
```diff
@@ -611,9 +613,12 @@ public static UTF8String exec(final UTF8String srcString) {
...
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:621]: and naturally `StringTrimRight`

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:1]: of course, please update the trim tests after the fix

**Review by uros-db — CHANGES_REQUESTED** [review]
let's also add some e2e sql tests to `collations.sql`

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 7 above)
Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java]: Should be fixed, thanks.

**Comment by jovanpavl-db** [comment]
> let's also add some e2e sql tests to `collations.sql`

done.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: please add a similar comment for `StringReplace`, and any other expressions in this class that you might've missed

**Comment by MaxGekk** [comment]
@uros-db @stefankandic @stevomitric @vladanvasi-db Could you take a look at the changes, please.

**Comment by jovanpavl-db** [comment]
> @uros-db @stefankandic @stevomitric @vladanvasi-db Could you take a look at the changes, please.

Synced offline with @uros-db, we decided to go with a bit different implementation of trim function. Will push changes soon and it will again be ready for the review. Sorry for not setting pr as WIP.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 10]
```diff
@@ -1191,6 +1204,17 @@ public static String[] getICULocaleNames() {
...
```
Comment (vladanvasi-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:1209]: Can we add a short comment for this since it is a public method?

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 11]
```diff
@@ -610,14 +625,19 @@ public static UTF8String exec(final UTF8String srcString) {
       return execBinary(srcString);
     }
     public static UTF8String exec(
-        final UTF8String srcString,
-        final UTF8String trimString,
-        final int collationId) {
+            final UTF8String srcString,
+            final UTF8String trimString,
+            final int collationId) {
```
Comment (vladanvasi-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:630]: nit: fix identation here

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 12]
```diff
@@ -932,7 +950,11 @@ case class StringReplace(srcExpr: Expression, searchExpr: Expression, replaceExp
 
   override def dataType: DataType = srcExpr.dataType
   override def inputTypes: Seq[AbstractDataType] =
-    Seq(StringTypeNonCSAICollation, StringTypeNonCSAICollation, StringTypeNonCSAICollation)
+    Seq(
+      StringTypeNonCSAICollation(supportsTrimCollation = true),
+      StringTypeNonCSAICollation(supportsTrimCollation = true),
+      StringTypeNonCSAICollation(supportsTrimCollation = true)
+    )
```
Comment (vladanvasi-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:957]: nit: Move this ) to the line above

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 13]
```diff
@@ -1232,22 +1255,62 @@ public static UTF8String trimLeft(
...
+   * @param srcString the input string to be trimmed from the right end of the string
+   * @param trimString the trim string characters to trim
+   * @param collationId the collation ID to use for string trim
+   * @return the trimmed string (for UTF8_LCASE collation)
+   */
+  public static UTF8String binaryTrimRight(
+      final UTF8String srcString,
+      final UTF8String trimString,
+      final int collationId) {
+    boolean ignoreSpaces = !trimString.contains(UTF8String.SPACE_UTF8) &&
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:1275]: not a huge fan of calling `contains` in O(n^2) here...

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 14]
```diff
@@ -1232,22 +1255,62 @@ public static UTF8String trimLeft(
...
+    while (lastNonSpacePosition > 0 &&
+            srcString.getByte(lastNonSpacePosition - 1) == 0x20) {
+      --lastNonSpacePosition;
+    }
+
+    StringBuilder sb = new StringBuilder();
+    sb.append(srcString.trimRight().trimRight(trimString));
+    if (lastNonSpacePosition != srcString.numChars()) {
+      sb.append(srcString.substring(lastNonSpacePosition, srcString.numChars()));
+    }
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:1291]: so here we have
- 2 UTF8String.trim calls, which will all allocate new UTF8Strings by calling `copyUTF8String`
- StringBuilder allocation
- additional UTF8String.substring call, which will allocate yet another UTF8String
- returning the final result, which first allocates a java String, and then another UTF8String

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:1291]: I think we should try to do better...

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: let's also add a comment to StringTranslate

please make sure to cover all expressions in CollationSupport - by either implementing space trimming handling, or adding a comment signaling that it's not required.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 15]
```diff
@@ -358,6 +358,19 @@ protected static UTF8String applyTrimmingPolicy(UTF8String s, int collationId) {
         return applyTrimmingPolicy(s, getSpaceTrimming(collationId));
       }
 
+      protected static boolean ignoresSpacesInTrimFunctions(
+              int collationId,
+              boolean isLTrim,
+              boolean isRTrim) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:364]: fix scalastyle

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 16]
```diff
@@ -1191,6 +1204,24 @@ public static String[] getICULocaleNames() {
...
+  }
+
+  /**
+   * Returns if leading/trailing spaces should be ignored in trim string expressions. This is needed
+   * because space trimming collation directly changes behaviour of trim functions.
+   */
+  public static boolean ignoresSpacesInTrimFunctions(
+          int collationId,
+          boolean isLTrim,
+          boolean isRTrim) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:1221]: fix scalastyle

**Review by uros-db — CHANGES_REQUESTED** [review]
couple small fixes, but one major thing we should consider is a more efficient implementation for `binaryTrim` - in general, let's try to:

- minimize extra allocations as much as we can
- let's use better trim chars querying (e.g. hashmap)
- try to build the string in one pass

for example, see the existing `trim` implementations in CollationAwareUTF8String
  public static UTF8String trimRight(
  public static UTF8String trimLeft(

**Comment by jovanpavl-db** [comment]
> couple small fixes, but one major thing we should consider is a more efficient implementation for `binaryTrim` - in general, let's try to:
> 
> * minimize extra allocations as much as we can
> * let's use better trim chars querying (e.g. hashmap)
> * try to build the string in one pass
> 
> for example, see the existing `trim` implementations in CollationAwareUTF8String public static UTF8String trimRight( public static UTF8String trimLeft(

I refactored a bit this both binaryTrim, lowercaseTrim  function to get the most optimal complexity. Here is a brief analysis (and also sanity check for me).

**binaryTrim**:
1. Used hash set to get on average O(1) comparison when searching trimString codes in srcString. Note that this is more efficient than original trim implementation in UTF8String  because there is linear search done, which brings complexity to O(len(srcString) * len(trimString)). 
2. Didn't knew that contains is O(n^2), used contain and ascii space code to avoid it and it boils down checks if spaces should be ignored to O(1).
3. At the end we have :`
    return UTF8String.concat(
      srcString.substring(0, searchIndex),
      srcString.substring(lastNonSpacePosition, srcString.numChars()));`
      this might look suspicious and true we have 1 copy-ing of O(len(searchIndex)) and 1 copying of O(number_trailing_spaces) and again copying those 2 (when concatenating), which is in the worst case O(2 * len(srcString)) O(srcString). So we did at the end get constant factor of 2, and true we can avoid it. But bottom line even in UTF8String rtrim implementation this copy-ing is done multiple time. On the other hand to avoid this 1-liner you need to literally reimplementing copyUTF8String(start1, end1, start2, end2) which I find ultra messy and frankly unnecessary, I'm really against it. 

**lcaseTrim**
1. Original code is minimally changed (not adding even constant factor). Point 3. from binaryTrim holds here also.
**

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 17]
```diff
@@ -358,6 +358,19 @@ protected static UTF8String applyTrimmingPolicy(UTF8String s, int collationId) {
         return applyTrimmingPolicy(s, getSpaceTrimming(collationId));
       }
 
+      protected static boolean ignoresSpacesInTrimFunctions(
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:361]: let's add a comment here explaining what this does / why we need it

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 18]
```diff
@@ -1357,8 +1514,16 @@ public static UTF8String trimRight(
...
+    // Return the substring from the start of the string until that position and append
+    // trailing spaces if they were ignored
+    int trailingLength = src.length() - lastNonSpacePosition;
+    StringBuilder sb =
+      new StringBuilder(charIndex + (Math.max(trailingLength, 0)));
+    sb.append(src, 0, charIndex);
+    if (lastNonSpacePosition != src.length()) {
+      sb.append(src, lastNonSpacePosition, src.length());
+    }
+    return UTF8String.fromString(sb.toString());
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:1526]: let's use the concat approach

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 19]
```diff
@@ -1329,7 +1467,26 @@ public static UTF8String trimRight(
...
     CharacterIterator target = new StringCharacterIterator(src);
     Collator collator = CollationFactory.fetchCollation(collationId).collator;
-    int charIndex = src.length(), longestMatchLen;
+    int charIndex = src.length(), longestMatchLen, lastNonSpacePosition = src.length();
+
+    // In cases of trim collation (rtrim for example) trailing spaces should be ignored.
+    // If trimString contains spaces this behaviour is not important as they would get trimmed
+    // anyway. However, if it is not the case they should be ignored and then appended after
+    // trimming other characters.
+    if (!trimChars.containsKey(0x20) &&
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:1476]: let's separate out 0x20 an (int) constant

**Review by uros-db — APPROVED** [review]
small comments left, otherwise lgtm

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 20]
```diff
@@ -1260,7 +1364,32 @@ public static UTF8String lowercaseTrimRight(
 
     // Iterate over `srcString` from the right to find the first character that is not in the set.
     int searchIndex = srcString.numChars(), codePoint, codePointBuffer = -1;
+
+    // In cases of trim collation (rtrim for example) trailing spaces should be ignored.
+    // If trimString contains spaces this behaviour is not important as they would get trimmed
+    // anyway. However, if it is not the case they should be ignored and then appended after
+    // trimming other characters.
+    int lastNonSpaceByteIdx = srcString.numBytes(), lastNonSpaceCharacterIdx = srcString.numChars();
+    if  (!trimChars.contains(0x20) &&
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:1373]:
```suggestion
    if (!trimChars.contains(0x20) &&
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 21]
```diff
@@ -1260,7 +1364,32 @@ public static UTF8String lowercaseTrimRight(
...
+      // In case of src string contains only spaces there is no need to do any trimming, since it's
+      // already checked that trim string does not contain any spaces.
+      if (lastNonSpaceByteIdx == 0) {
+        return srcString;
+      }
+      searchIndex = lastNonSpaceCharacterIdx =
+        srcString.numChars() - (srcString.numBytes() - lastNonSpaceByteIdx);
+    }
     Iterator<Integer> srcIter = srcString.reverseCodePointIterator();
+    for(int i = lastNonSpaceCharacterIdx; i < srcString.numChars(); i++) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:1389]:
```suggestion
    for (int i = lastNonSpaceCharacterIdx; i < srcString.numChars(); i++) {
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 22]
```diff
@@ -1232,22 +1255,103 @@ public static UTF8String trimLeft(
...
+      // In case of src string contains only spaces there is no need to do any trimming, since it's
+      // already checked that trim string does not contain any spaces.
+      if (lastNonSpaceByteIdx == 0) {
+        return srcString;
+      }
+      searchIndex = lastNonSpaceCharacterIdx =
+        srcString.numChars() - (srcString.numBytes() - lastNonSpaceByteIdx);
+    }
+    Iterator<Integer> srcIter = srcString.reverseCodePointIterator();
+    for(int i = lastNonSpaceCharacterIdx; i < srcString.numChars(); i++) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:1309]:
```suggestion
    for (int i = lastNonSpaceCharacterIdx; i < srcString.numChars(); i++) {
```

**Comment by uros-db** [comment]
also, look out for any scalastyle issues

**Comment by jovanpavl-db** [comment]
@MaxGekk can you please take a look?

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 23]
```diff
@@ -539,26 +549,35 @@ public static UTF8String execBinary(
...
     public static UTF8String execICU(
         final UTF8String srcString,
         final UTF8String trimString,
         final int collationId) {
       return CollationAwareUTF8String.trim(srcString, trimString, collationId);
     }
+    public static UTF8String execBinaryTrim(
+            final UTF8String srcString,
+            final UTF8String trimString,
+            final int collationId) {
```
Comment (MaxGekk) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:565]: please, fix indentations here

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 24]
```diff
@@ -1080,6 +1080,24 @@ public static UTF8String translate(final UTF8String input,
...
+  /**
+   * Trims the `srcString` string from both ends of the string using the specified `trimString`
+   * characters, with respect to the UTF8_BINARY trim collation. String trimming is performed by
+   * first trimming the left side of the string, and then trimming the right side of the string.
+   * The method returns the trimmed string. If the `trimString` is null, the method returns null.
+   *
+   * @param srcString the input string to be trimmed from both ends of the string
+   * @param trimString the trim string characters to trim
+   * @param collationId the collation ID to use for string trim
+   * @return the trimmed string (for UTF8_LCASE collation)
```
Comment (MaxGekk) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:1092]: I didn't get why do you highlight `UTF8_LCASE`. Doesn't the function return a trimmed string for other collations like `UTF8_BINARY`?

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 25]
```diff
@@ -1232,22 +1255,103 @@ public static UTF8String trimLeft(
...
+   * with respect to the UTF8_BINARY trim collation. For UTF8_BINARY trim collation, the method has
+   * one special case to cover with respect to trimRight function for regular UTF8_Binary collation.
+   * Trailing spaces should be ignored in case of trim collation (rtrim for example) and if
+   * trimString does not contain spaces. In this case, the method trims the string from the right
+   * and after call of regular trim functions returns back trimmed spaces as those should not get
+   * removed.
+   * @param srcString the input string to be trimmed from the right end of the string
+   * @param trimString the trim string characters to trim
+   * @param collationId the collation ID to use for string trim
+   * @return the trimmed string (for UTF8_LCASE collation)
```
Comment (MaxGekk) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:1269]: and for non-UTF8_LCASE, what does it return?

**Review by MaxGekk — COMMENTED** [review]
@jovanpavl-db Could you resolve conflicts, please. In general, LGTM.

**Comment by jovanpavl-db** [comment]
> @jovanpavl-db Could you resolve conflicts, please. In general, LGTM.

Done.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @jovanpavl-db and @uros-db @vladanvasi-db for review.
